src/utils.ts

```diff
@@ -28,3 +28,49 @@ export async function getCmdOutput(
   }
   return stdout;
 }
+
+export async function withRetries<T>(
+  operation: () => Promise<T>,
+  maxRetryAttempts: number,
+  isRetriable: (error: unknown) => boolean
+): Promise<T> {
+  let attemptsLeft = maxRetryAttempts;
+  while (true) {
+    try {
+      return await operation();
+    } catch (e: unknown) {
+      attemptsLeft -= 1;
+      if (attemptsLeft <= 0) {
+        throw e;
+      }
+      if (!isRetriable(e)) {
+        throw e;
+      }
+      core.info(
+        `[warning] Retrying after an error, ${attemptsLeft} attempts left, error: ${e}`
+      );
+    }
+  }
+}
+
+class TimeoutError extends Error {}
+
+export async function withTimeout<T>(
+  operation: (onTimeout: Promise<void>) => Promise<T>,
+  timeoutMs: null | number
+): Promise<T> {
+  const timeout = timeoutMs
+    ? new Promise<void>((resolve) => {
+        setTimeout(resolve, timeoutMs);
+      })
+    : new Promise<never>(() => {});
+
+  const timeoutSym = Symbol("timeout" as const);
+  const racingTimeout = timeout.then(() => timeoutSym);
```

Comment:
```suggestion
  const racingTimeout = timeout.then(() => { throw new TimeoutError("operation timeout"); });
```

This way you shouldn’t need the symbol and explicit check, just `return await Promise.race`